native/dlahqr.go

```diff
@@ -0,0 +1,418 @@
+// sqrt(abs(H[i+1,i]*H[i,i+1])) and wi[i+1] = -wi[i].
+//
+// wr and wi must have length n.
+//
+// z and ldz represent an n×n matrix Z. If wantz is true, the transformations
+// will be applied to the submatrix Z[iloz:ihiz+1,ilo:ihi+1]. If wantz is false,
+// z is not referenced. It must hold that
+//  0 <= iloz <= ilo; ihi <= ihiz < n.
+//
+// If info == 0 and wantt is true, on return H[ilo:ihi+1,ilo:ihi+1] will be
+// overwritten by upper quasi-triangular full Schur form with any 2×2 diagonal
+// blocks in standard form.
+//
+// If info == 0 and wantt is false, the contents of h on return is unspecified.
+//
+// If info > 0, Dlahqr failed to compute all the eigenvalues ilo to ihi in a
+// total of 30 iterations per eigenvalue. wr[info+1:ihi+1] and wi[info+1:ihi+1]
+// contain those eigenvalues which have been successfully computed.
+//
+// If info > 0 and wantt is false, on return, the remaining unconverged
+// eigenvalues are the eigenvalues of the upper Hessenberg matrix
+// H[ilo:info+1,ilo:info+1].
+//
+// If info > 0 and wantt is true, then on return
+//  (initial H)*U = U*(final H),   (*)
+// where U is an orthogonal matrix. The final H is upper Hessenberg and
+// H[info+1:ihi+1,info+1:ihi+1] is upper triangular.
+//
+// If info > 0 and wantz is true, then on return
+//  (final Z) = (initial Z)*U,
+// where U is the orthogonal matrix in (*) regardless of the value of wantt.
+//
+// Dlahqr is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dlahqr(wantt, wantz bool, n, ilo, ihi int, h []float64, ldh int, wr, wi []float64, iloz, ihiz int, z []float64, ldz int) (info int) {
+	checkMatrix(n, n, h, ldh)
+	switch {
+	case ilo < 0 || max(0, ihi) < ilo:
+		panic(badIlo)
+	case n <= ihi:
+		panic(badIhi)
+	case len(wr) != n:
+		panic("lapack: bad length of wr")
+	case len(wi) != n:
+		panic("lapack: bad length of wi")
+	case ilo > 0 && h[ilo*ldh+ilo-1] != 0:
+		panic("lapack: block is not isolated")
+	}
+	if wantz {
+		checkMatrix(n, n, z, ldz)
+		switch {
+		case iloz < 0 || ilo < iloz:
+			panic("lapack: iloz out of range")
+		case ihiz < ihi || n <= ihiz:
+			panic("lapack: ihiz out of range")
+		}
+	}
+
+	// Quick return if possible.
+	if n == 0 {
+		return info
+	}
+	if ilo == ihi {
+		wr[ilo] = h[ilo*ldh+ilo]
+		wi[ilo] = 0
+		return info
+	}
+
+	// Clear out the trash.
+	for j := ilo; j < ihi-2; j++ {
+		h[(j+2)*ldh+j] = 0
+		h[(j+3)*ldh+j] = 0
+	}
+	if ilo <= ihi-2 {
+		h[ihi*ldh+ihi-2] = 0
+	}
+
+	nh := ihi - ilo + 1
+	nz := ihiz - iloz + 1
+
+	// Set machine-dependent constants for the stopping criterion.
+	ulp := dlamchP
+	smlnum := float64(nh) / ulp * dlamchS
+
+	// i1 and i2 are the indices of the first row and last column of H to
+	// which transformations must be applied. If eigenvalues only are being
+	// computed, i1 and i2 are set inside the main loop.
+	var i1, i2 int
+	if wantt {
+		i1 = 0
+		i2 = n - 1
+	}
+
+	itmax := 30 * max(10, nh) // Total number of QR iterations allowed.
```

Comment:
This is actually not true (now) since the for loop was changed. I don't know why the Fortran does what it does.